src/pybind/mgr/dashboard/controllers/docs.py

```diff
@@ -451,3 +448,35 @@ def __call__(self, all_endpoints=False):
               query_params="{all_endpoints}")
     def _with_token(self, token, all_endpoints=False):
         return self._swagger_ui_page(all_endpoints, token)
+
+
+if __name__ == "__main__":
+    import sys
+    import os
+    import yaml
+    from . import generate_routes
+# a hot fix for errors caused by null description values when generating
+# static documentation: better fix would be default values in source
+# to be 'None' strings: however, decorator changes didn't resolve
+
+    def fix_null_descr(api_json):
+        fixed = api_json
+        for k, v in api_json.items():
+            if isinstance(v, dict):
+                fix_null_descr(v)
+            elif v is None:
+                fixed[k] = 'None'
+        return fixed
+
+    mgr.get_frontend_path.side_effect = lambda: os.path.abspath('./dashboard/frontend/dist')
+    generate_routes("/api")
+    try: 
+        with open(sys.argv[1], 'w') as f:
+            # pylint: disable=protected-access
+            yaml.dump(
+                fix_null_descr(Docs._gen_spec(all_endpoints=False, base_url="/", offline=True)),
+                f)
+    except IndexError as ind_err:
+        print(ind_err)
+        print("ERROR: It appears no output file was specified.")
+        raise Exception("Output file name missing; correct syntax is: `tox -e apidocs <file>.yml`")
```

Comment:
If I run `tox -e apidocs dash_local.yml` in `src/pybind/mgr/dashboard` it won't create `dash_local.yml` in `src/pybind/mgr/dashboard` but in `src/pybind/mgr`. As this feels odd I needed `find` to find it as it wasn't in the predicted directory. If I use an absolute path it works but not with a relative one.

I assume it's caused by `changedir` in `tox.ini`. It's fine for me if we only support absolutes paths here, but please note that anywhere as the current exception says it's supporting relative filepaths.

Reply:
Yeah, for invoking the docs controller as a module, this needs to travel the path one level up... Not sure if there's an easy way in Python to work around this issue...